packages/api/cms-api/src/dam/files/files.controller.ts

```diff
@@ -9,136 +9,163 @@ import {
+    @Controller("dam/files")
+    class FilesController {
+        constructor(
+            @Inject(DAM_CONFIG) private readonly damConfig: DamConfig,
+            private readonly filesService: FilesService,
+            private readonly blobStorageBackendService: BlobStorageBackendService,
+        ) {}
 
-    private checkCdnOrigin(incomingCdnOriginHeader: string): void {
-        if (this.damConfig.cdnEnabled) {
-            if (incomingCdnOriginHeader !== this.damConfig.cdnOriginHeader) {
-                throw new ForbiddenException();
+        @Post("upload")
+        @UseInterceptors(DamUploadFileInterceptor(FilesService.UPLOAD_FIELD))
+        async upload(@UploadedFile() file: FileUploadInterface, @Body() body: UploadFileBodyInterface): Promise<FileInterface> {
+            const transformedBody = plainToInstance(UploadFileBody, body);
+            const errors = await validate(transformedBody, { whitelist: true, forbidNonWhitelisted: true });
+
+            if (errors.length > 0) {
+                throw new CometValidationException("Validation failed", errors);
             }
+
+            const { scope, folderId } = transformedBody;
+            const uploadedFile = await this.filesService.upload({ file, folderId }, nonEmptyScopeOrNothing(scope));
+            return Object.assign(uploadedFile, { fileUrl: await this.filesService.createFileUrl(uploadedFile) });
         }
-    }
 
-    private isValidHash(hash: string, fileParams: FileParams): boolean {
-        return hash === this.filesService.createHash(fileParams);
-    }
+        @Get(`/preview/${fileUrl}`)
+        async previewFileUrl(@Param() params: FileParams, @Res() res: Response, @Headers("range") range?: string): Promise<void> {
+            return this.streamFile(params, res, { range, overrideHeaders: { "Cache-control": "private" } });
+        }
 
-    private async streamFile(
-        { fileId }: FileParams,
-        res: Response,
-        options?: {
-            range?: string;
-            overrideHeaders?: OutgoingHttpHeaders;
-        },
-    ): Promise<void> {
-        const file = await this.filesService.findOneById(fileId);
-        if (!file) throw new NotFoundException();
-
-        const headers = {
-            "content-type": file.mimetype,
-            "last-modified": file.updatedAt?.toUTCString(),
-            "content-length": file.size,
-        };
-
-        // https://medium.com/@vishal1909/how-to-handle-partial-content-in-node-js-8b0a5aea216
-        let response: NodeJS.ReadableStream;
-        if (options?.range) {
-            const { start, end, contentLength } = calculatePartialRanges(file.size, options.range);
-
-            if (start >= file.size || end >= file.size) {
-                res.writeHead(416, {
-                    "content-range": `bytes */${file.size}`,
-                });
-                res.end();
-                return;
+        @DisableGlobalGuard()
+        @Get(`/:hash/${fileUrl}`)
+        async hashedFileUrl(
```

Comment:
btw: how is this made secure? Just by hashing? does this url have a timeout?

Reply:
As far as I know, only by hashing. I'm not aware that we have an item on our backlog to change it. We should discuss it internally